mixer/adapter/kubernetes2/kubernetes.go

```diff
@@ -0,0 +1,409 @@
+var _ kubernetes_apa_tmpl.HandlerBuilder = &builder{}
+
+/////////////////////////////// Config-time methods //////////////////////////////
+//
+//
+
+// GetInfo returns the Info associated with this adapter implementation.
+func GetInfo() adapter.Info {
+	return adapter.Info{
+		Name:        "kubernetesenv",
+		Impl:        "istio.io/istio/mixer/adapter/kubernetes",
+		Description: "Provides platform specific functionality for the kubernetes environment",
+		SupportedTemplates: []string{
+			kubernetes_apa_tmpl.TemplateName,
+		},
+		DefaultConfig: conf,
+
+		NewBuilder: func() adapter.HandlerBuilder { return newBuilder(newCacheFromConfig) },
+	}
+}
+
+func (b *builder) SetAdapterConfig(c adapter.Config) {
+	b.adapterConfig = c.(*config.Params)
+}
+
+// Validate is responsible for ensuring that all the configuration state given to the builder is
+// correct.
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	params := b.adapterConfig
+	if len(params.PodLabelForService) == 0 {
+		ce = ce.Appendf("podLabelForService", "field must be populated")
+	}
+	if len(params.PodLabelForIstioComponentService) == 0 {
+		ce = ce.Appendf("podLabelForIstioComponentService", "field must be populated")
+	}
+	if len(params.FullyQualifiedIstioIngressServiceName) == 0 {
+		ce = ce.Appendf("fullyQualifiedIstioIngressServiceName", "field must be populated")
+	}
+	if len(params.ClusterDomainName) == 0 {
+		ce = ce.Appendf("clusterDomainName", "field must be populated")
+	} else if len(strings.Split(params.ClusterDomainName, ".")) != 3 {
+		ce = ce.Appendf("clusterDomainName", "must have three segments, separated by '.' ('svc.cluster.local', for example)")
+	}
+	return
+}
+
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+	paramsProto := b.adapterConfig
+	stopChan := make(chan struct{})
+	refresh := paramsProto.CacheRefreshDuration
+	path, exists := os.LookupEnv("KUBECONFIG")
+	if !exists {
+		path = paramsProto.KubeconfigPath
+	}
+	controller, err := b.newCacheControllerFn(path, refresh, env)
+	if err != nil {
+		return nil, err
+	}
+	env.ScheduleDaemon(func() { controller.Run(stopChan) })
+	// ensure that any request is only handled after
+	// a sync has occurred
+	if env.Logger().VerbosityLevel(debugVerbosityLevel) {
+		env.Logger().Infof("Waiting for kubernetes cache sync...")
+	}
+	if success := cache.WaitForCacheSync(stopChan, controller.HasSynced); !success {
+		stopChan <- struct{}{}
+		return nil, errors.New("cache sync failure")
+	}
+	if env.Logger().VerbosityLevel(debugVerbosityLevel) {
+		env.Logger().Infof("Cache sync successful.")
+	}
+	return &handler{
+		log:    env.Logger(),
+		pods:   controller,
+		params: paramsProto,
+	}, nil
+}
+
+func newBuilder(cacheFactory controllerFactoryFn) *builder {
+	return &builder{
+		newCacheControllerFn: cacheFactory,
+		adapterConfig:        conf,
+	}
+}
+
+/////////////////////////////// Request-time methods //////////////////////////////
+//
```

Comment:
what are these extra lines of `//` needed for?

Reply:
Removed